From a7f38ce454236ac9809181e2d6656ea6fd312f93 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Dominik=20R=C3=B6ttsches?= <drott@chromium.org>
Date: Mon, 19 Jan 2026 15:40:39 +0200
Subject: [PATCH] [xml] Patch xml crate to include performance improvements

Imports https://github.com/kornelski/xml-rs/pull/62 while
there is no upstream tagged release including that change yet.
---
 .../vendor/xml-v1/src/reader/indexset.rs               |  2 +-
 .../vendor/xml-v1/src/reader/parser.rs                 | 10 ++++++----
 2 files changed, 7 insertions(+), 5 deletions(-)

diff --git a/third_party/rust/chromium_crates_io/vendor/xml-v1/src/reader/indexset.rs b/third_party/rust/chromium_crates_io/vendor/xml-v1/src/reader/indexset.rs
index bb5069c92f3a6..af91597422732 100644
--- a/third_party/rust/chromium_crates_io/vendor/xml-v1/src/reader/indexset.rs
+++ b/third_party/rust/chromium_crates_io/vendor/xml-v1/src/reader/indexset.rs
@@ -21,7 +21,7 @@ const HASH_THRESHOLD: usize = 8;
 impl AttributesSet {
     pub fn new() -> Self {
         Self {
-            vec: Vec::new(),
+            vec: Vec::with_capacity(HASH_THRESHOLD),
             hasher: RandomState::new(),
             may_contain: HashSet::default(),
         }
diff --git a/third_party/rust/chromium_crates_io/vendor/xml-v1/src/reader/parser.rs b/third_party/rust/chromium_crates_io/vendor/xml-v1/src/reader/parser.rs
index e60db8bd06614..e8a9967fdbf8c 100644
--- a/third_party/rust/chromium_crates_io/vendor/xml-v1/src/reader/parser.rs
+++ b/third_party/rust/chromium_crates_io/vendor/xml-v1/src/reader/parser.rs
@@ -15,6 +15,8 @@ use crate::reader::lexer::{Lexer, Token};
 use std::collections::HashMap;
 use std::io::Read;
 
+static STRING_RESERVE_CAPACITY: usize = 20;
+
 macro_rules! gen_takes(
     ($($field:ident -> $method:ident, $t:ty, $def:expr);+) => (
         $(
@@ -31,8 +33,8 @@ macro_rules! gen_takes(
 );
 
 gen_takes!(
-    name         -> take_name, String, String::new();
-    ref_data     -> take_ref_data, String, String::new();
+    name         -> take_name, String, String::with_capacity(STRING_RESERVE_CAPACITY);
+    ref_data     -> take_ref_data, String, String::with_capacity(STRING_RESERVE_CAPACITY);
 
     encoding     -> take_encoding, Option<String>, None;
 
@@ -118,7 +120,7 @@ impl PullParser {
             lexer,
             st: State::DocumentStart,
             state_after_reference: State::OutsideTag,
-            buf: String::new(),
+            buf: String::with_capacity(STRING_RESERVE_CAPACITY),
             entities: HashMap::new(),
             nst: NamespaceStack::default(),
 
@@ -500,7 +502,7 @@ impl PullParser {
 
     #[inline]
     fn take_buf(&mut self) -> String {
-        std::mem::take(&mut self.buf)
+        std::mem::replace(&mut self.buf, String::with_capacity(STRING_RESERVE_CAPACITY))
     }
 
     #[inline]
-- 
2.52.0.457.g6b5491de43-goog

